build.gradle

```diff
@@ -463,3 +463,46 @@ def configureGroovydoc(TaskProvider groovydoc) {
     include "spock/**"
   }
 }
+
+configurations {
+  kotlinCompilerClasspath {
+    canBeConsumed = false
+  }
+  kotlinScriptClasspath {
+    canBeConsumed = false
+  }
+}
+
+dependencies {
+  kotlinCompilerClasspath 'org.jetbrains.kotlin:kotlin-compiler:1.8.20'
+  kotlinCompilerClasspath 'org.jetbrains.kotlin:kotlin-scripting-compiler:1.8.20'
+  kotlinScriptClasspath('org.jetbrains.kotlin:kotlin-main-kts:1.8.20') { transitive = false }
+}
+
+def preprocessWorkflows = tasks.register('preprocessWorkflows')
+file('.github/workflows').eachFileMatch(~/.*\.main\.kts$/) { workflowScript ->
+  def workflowName = workflowScript.name - ~/\.main\.kts$/
+  def pascalCasedWorkflowName = workflowName
+    .replaceAll(/-\w/) { it[1].toUpperCase() }
+    .replaceFirst(/^\w/) { it[0].toUpperCase() }
+  def preprocessWorkflow = tasks.register("preprocess${pascalCasedWorkflowName}Workflow", JavaExec) {
+    inputs
+      .file(workflowScript)
+      .withPropertyName('workflowScript')
+    outputs
+      .file(new File(workflowScript.parent, "${workflowName}.yaml"))
+      .withPropertyName('workflowFile')
+
+    javaLauncher = javaToolchains.launcherFor {
+      languageVersion = JavaLanguageVersion.of(17)
+    }
+    classpath(configurations.kotlinCompilerClasspath)
+    mainClass = 'org.jetbrains.kotlin.cli.jvm.K2JVMCompiler'
+    args('-no-stdlib', '-no-reflect')
+    args('-classpath', configurations.kotlinScriptClasspath.asPath)
+    args('-script', workflowScript.absolutePath)
+  }
+  preprocessWorkflows.configure {
+    dependsOn(preprocessWorkflow)
+  }
```

Comment:
This whole thing should be extracted into `build-logic` this files is already too bloated.

Reply:
Also can we add the validation here and just have on job that validates all workflows instead of having it inline of each job?

Reply:
> This whole thing should be extracted into build-logic this files is already too bloated.

Yep, no problem, moved to a `build-logic` subproject.

> Also can we add the validation here and just have on job that validates all workflows instead of having it inline of each job?

Well, yes and no.
Indeed we should have a check that verifies all YAMLs for PRs and branches as not all run always like the "release" one.
I added this now.
You can disable the per-workflow consistency check that is generated, but I would recommend to keep it there.
The YAML is basically just the output format of the Kotlin scripts and that it is there additionally should not be too much of a problem.
Until https://github.com/orgs/community/discussions/15904 is considered and realized, you just have to run the scripts manually and check in the result and thus need the consistency check.
So I'd suggest to keep the per-file consistency check there, as then the jobs defined in the respective file are also always only executed if the consistency check was successful.